{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "ImportError",
     "evalue": "attempted relative import with no known parent package",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-3b801d86d24a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# Import deSeq globals\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[1;33m.\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcode\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mglobals\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mBARCODE_LENGTH\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mADAPTER_LENGTH_F\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mADAPTER_LENGTH_R\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;31m# Import required modules\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mImportError\u001b[0m: attempted relative import with no known parent package"
     ]
    }
   ],
   "source": [
    "# Import deSeq globals\n",
    "from ..code.globals import BARCODE_LENGTH, ADAPTER_LENGTH_F, ADAPTER_LENGTH_R\n",
    "\n",
    "# Import required modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from string import Template\n",
    "from random import shuffle\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the allowed bases\n",
    "allowed_bases_tup = (\"A\", \"T\", \"C\", \"G\")\n",
    "allowed_bases_set = set(allowed_bases_tup)\n",
    "\n",
    "# Define potential \"in frame\" indices and allowed\n",
    "# indices for making mutations within a codon\n",
    "allowed_in_frame = (0, 1, 2)\n",
    "allowed_n_mutations = (1, 2, 3)\n",
    "\n",
    "# Define the frequency with which no mutations are made\n",
    "wt_freq = 0.05\n",
    "\n",
    "# Define the allowed lengths of reference sequences\n",
    "max_ref_length = 601\n",
    "\n",
    "# Define the allowed number of reads in a well\n",
    "allowed_n_reads = (0, 101)\n",
    "\n",
    "# Determine how many variants are allowed in a well\n",
    "allowed_n_variants = (1, 4)\n",
    "\n",
    "# Determine the frequency with which we have dead wells\n",
    "dead_well_freq = 0.05\n",
    "\n",
    "# Define the probability that any given base is bad\n",
    "bad_base_freq = 0.05\n",
    "\n",
    "# Load the barcode df\n",
    "barcode_df = pd.read_csv(\"./Code/IndexMap.csv\")\n",
    "\n",
    "# Define a forward ID template\n",
    "id_template = Template(\"@M06418:33:000000000-CRL6Y:1:1101:0:$num 1:N:0:162\")\n",
    "\n",
    "# Define the forward and reverse adapters\n",
    "F_ADAPTER = \"CACCCAAGACCACTCTCCGG\"\n",
    "R_ADAPTER = \"CGGTGTGCGAAGTAGGTGC\"\n",
    "\n",
    "# Define the frequencies with which we ruin forward and reverse reads with\n",
    "# different strategies\n",
    "indel_freq = 0.01\n",
    "low_qual_freq = 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is used to build test cases for the new deSeq pipeline. The main things to test are:\n",
    "1. The outputs of files are as we expect.\n",
    "2. Everything works stably.\n",
    "\n",
    "We will build both general test conditions as well as make a stress test which randomly generates input conditions and records instances of failure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Conditions in particular to look at:\n",
    "1. How well does deSeq handle (0) imperfectly overlapping reads (e.g. reads go into the barcode regions) (1) perfectly overlapping reads (2) overlapping reads (3) perfectly non-overlapping (where the forward ends just before reverse starts) and (4) non-overlapping.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a given well and set of deSeq parameters...\n",
    "1. Randomly decide which positions are mutated\n",
    "2. Randomly decide on a count and frequency for the full combination\n",
    "3. Decide if the well is wild type (if the combination is wild type) or dead (if the count is below our expected frequency)\n",
    "4. Build reads with quality scores, lengths, etc. that will generate the target output for the combination\n",
    "    1. Consider all quality filters at this stage! The final product should be a combo of all of them so that we can test everything together.\n",
    "    2. Make sure you mix up whether just the forward read is bad, just the reverse read is bad, etc.\n",
    "5. Count the expected individual read counts and calculate a frequency \n",
    "6. Spike in random reads with insertions and deletions across the sequence\n",
    "\n",
    "Repeat the above for a series of wells to make up a fake dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define run-specific parameters\n",
    "average_q_cutoff = 25\n",
    "bp_q_cutoff = 30\n",
    "variable_count = 10\n",
    "read_length = 150\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def build_mutant_sequence():\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_codon(parent_codon, mut_inds, allowed_mutations):\n",
    "    \n",
    "    # Define the new codon\n",
    "    new_codon = list(parent_codon)\n",
    "    new_allowed = [None] * len(mut_inds)\n",
    "    \n",
    "    # Loop over mutated positions\n",
    "    for i, ind in enumerate(mut_inds):\n",
    "\n",
    "        # Choose the new base\n",
    "        new_base = np.random.choice(tuple(allowed_mutations[i]))\n",
    "        \n",
    "        # Update the codon\n",
    "        new_codon[ind] = new_base\n",
    "        \n",
    "        # Update the allowed mutations\n",
    "        new_allowed[i] = allowed_mutations[i] - set(new_base)\n",
    "\n",
    "    # Return the new codon joined as a string and the updated\n",
    "    # set of allowed codons\n",
    "    return \"\".join(new_codon), new_allowed\n",
    "\n",
    "# Write functions that break up the larger procedure below\n",
    "def find_start_end_muts(nub_seq, nub_len):\n",
    "    \n",
    "    # Get the number of mutations in the starting and ending sequence\n",
    "    nub_inds = np.arange(nub_len)\n",
    "    n_nub_muts = np.random.choice(nub_inds + 1)\n",
    "\n",
    "    # Get the positions of mutations in the starting and ending sequence\n",
    "    nub_mut_pos = np.sort(np.random.choice(nub_inds, size = n_nub_muts, replace = False))\n",
    "\n",
    "    # Get the allowed starting mutations\n",
    "    allowed_nub_muts = [allowed_bases_set - set(nub_seq[mut_pos]) for mut_pos in nub_mut_pos]\n",
    "    \n",
    "    # Return needed info\n",
    "    return nub_mut_pos, allowed_nub_muts    \n",
    "\n",
    "class WellContents():\n",
    "    \n",
    "    # Initialize\n",
    "    def __init__(self, ref_seq = None, ref_length = None, read_length = 150):\n",
    "        \n",
    "        # Randomly generate a reference sequence\n",
    "        if ref_length == None:\n",
    "            self.ref_length = np.random.randint(read_length - 26, max_ref_length)\n",
    "        else:\n",
    "            self.ref_length = ref_length\n",
    "        if ref_seq == None:\n",
    "            self.ref_seq = \"\".join(np.random.choice(allowed_bases_tup, size = self.ref_length))\n",
    "        else:\n",
    "            self.ref_seq = ref_seq\n",
    "\n",
    "        # Assign what is in frame and split into codons\n",
    "        self.in_frame = np.random.choice(allowed_in_frame)\n",
    "        chunk_inds = [np.arange(i, i+3) for i in range(self.in_frame, self.ref_length, 3)]\n",
    "        all_chunks = [self.ref_seq[i:i+3] for i in range(self.in_frame, self.ref_length, 3)]\n",
    "        self.codon_bp_inds = [chunk for chunk in chunk_inds if len(chunk) == 3]\n",
    "        self.all_codons = [chunk for chunk in all_chunks if len(chunk) == 3]\n",
    "        self.n_codons = len(self.all_codons)\n",
    "        self.codon_inds = np.arange(self.n_codons)\n",
    "            \n",
    "        # Get the amino acid sequences\n",
    "        joined_readable = \"\".join(self.all_codons)\n",
    "        self.aa_ref_seq = str(Seq(joined_readable).translate())\n",
    "\n",
    "        # Map basepair to codon\n",
    "        readable_bases = len(joined_readable)\n",
    "        self.bp_to_codon = {ind: (ind - self.in_frame)//3 for ind in range(self.in_frame, self.in_frame + readable_bases)}\n",
    "        \n",
    "        # Map codon to basepair\n",
    "        self.codon_to_bp = {}\n",
    "        bp_counter = self.in_frame\n",
    "        for codon_ind, codon in enumerate(self.all_codons):\n",
    "            self.codon_to_bp[codon_ind] = np.array([bp_counter, bp_counter + 1, bp_counter + 2])\n",
    "            bp_counter += 3\n",
    "        \n",
    "        # Find the starting and ending sequence nubs\n",
    "        self.start_seq = self.ref_seq[:self.in_frame]\n",
    "        self.start_len = len(self.start_seq)\n",
    "        self.end_seq = all_chunks[-1] if len(all_chunks[-1]) != 3 else \"\"\n",
    "        self.end_len = len(self.end_seq)\n",
    "\n",
    "        # Build variants and make the mutant sequences\n",
    "        self.build_variants()\n",
    "        self.build_mutant_seqs()\n",
    "        \n",
    "        # Get the count array for a perfect sequence\n",
    "        self.read_length = read_length\n",
    "        self.get_readable_regions()\n",
    "        self.get_perfect_count_array()\n",
    "        \n",
    "        # Find the variable positions\n",
    "        self.find_variable_positions()\n",
    "                \n",
    "    def build_variants(self):\n",
    "        \n",
    "        # Determine how many variants you want in a well\n",
    "        self.n_variants = np.random.randint(*allowed_n_variants)\n",
    "\n",
    "        # Choose the number of mutated positions, the number of mutations\n",
    "        # in each codon, and where in the codon the mutation takes place\n",
    "        n_pos_mutated = np.random.randint(1, int(self.n_codons * 0.1)) # Total number of mutations\n",
    "        self.codon_mutation_positions = np.sort(np.random.choice(self.codon_inds, size = n_pos_mutated, replace = False)) # Codons to mutate\n",
    "        all_n_mutations = np.random.choice(allowed_n_mutations, size = n_pos_mutated) # Number of mutations at each codon\n",
    "        self.all_codon_mut_pos = [np.sort(np.random.choice(allowed_in_frame, size = n_mutations, replace = False))\n",
    "                                  for n_mutations in all_n_mutations]\n",
    "\n",
    "        # Create a list of allowed mutations at each position\n",
    "        allowed_mutations = [None] * n_pos_mutated\n",
    "        for i, mutation_pos in enumerate(self.codon_mutation_positions):\n",
    "            allowed_mutations[i] = [allowed_bases_set - set(self.all_codons[mutation_pos][mut_base_ind])\n",
    "                                   for mut_base_ind in self.all_codon_mut_pos[i]]\n",
    "\n",
    "        # Check to see if we want to mutate start positions\n",
    "        if self.start_len > 0:\n",
    "             self.start_mut_pos, allowed_start_muts = find_start_end_muts(self.start_seq, self.start_len)\n",
    "\n",
    "        # Check to see if we want to mutate end positions\n",
    "        if self.end_len > 0:\n",
    "            self.end_mut_pos, allowed_end_muts = find_start_end_muts(self.end_seq, self.end_len)\n",
    "\n",
    "        # Loop over the number of variants and generate new sequences\n",
    "        self.new_codons = [\"\"] * self.n_variants\n",
    "        self.new_starts = [\"\"] * self.n_variants\n",
    "        self.new_ends = [\"\"] * self.n_variants\n",
    "        for variant_ind in range(self.n_variants):\n",
    "\n",
    "            # Mutate all codons\n",
    "            new_codon_list = self.all_codons.copy()\n",
    "            for i, (mut_pos, codon_mut_pos, codon_allowed_mutations) in enumerate(zip(self.codon_mutation_positions,\n",
    "                                                                                      self.all_codon_mut_pos,\n",
    "                                                                                      allowed_mutations)):\n",
    "\n",
    "                # Mutate the codon\n",
    "                new_codon, new_allowed = mutate_codon(new_codon_list[mut_pos], codon_mut_pos, codon_allowed_mutations)\n",
    "\n",
    "                # Update the new sequence and new allowed\n",
    "                new_codon_list[mut_pos] = new_codon\n",
    "                allowed_mutations[i] = new_allowed\n",
    "\n",
    "                # Update the list of new sequences\n",
    "                self.new_codons[variant_ind] = new_codon_list\n",
    "\n",
    "            # Check to see if we want to mutate the starting sequence. We will only\n",
    "            # mutate if there is one\n",
    "            if self.start_len > 0:\n",
    "                self.new_starts[variant_ind], allowed_start_muts = mutate_codon(self.start_seq,\n",
    "                                                                                self.start_mut_pos,\n",
    "                                                                                allowed_start_muts)\n",
    "\n",
    "            # Check to see if we want to mutate the end sequence. We will only\n",
    "            # mutate if there is one\n",
    "            if self.end_len > 0:\n",
    "                self.new_ends[variant_ind], allowed_end_muts = mutate_codon(self.end_seq,\n",
    "                                                                            self.end_mut_pos,\n",
    "                                                                            allowed_end_muts)\n",
    "    \n",
    "    # Build sequences from the mutants\n",
    "    def build_mutant_seqs(self):\n",
    "        \n",
    "        # Define lists for holding the new sequences\n",
    "        self.mutant_aas = [None] * self.n_variants\n",
    "        self.mutant_seqs = [None] * self.n_variants\n",
    "        \n",
    "        # Loop over all mutants and append to new lists\n",
    "        for i, (start, mid, end) in enumerate(zip(self.new_starts, self.new_codons, self.new_ends)):\n",
    "\n",
    "            # Get the new translatable sequence\n",
    "            new_translatable = Seq(\"\".join(mid))\n",
    "            \n",
    "            # Complete the mutated sequence\n",
    "            self.mutant_seqs[i] = \"\".join([start, str(new_translatable), end])\n",
    "            \n",
    "            # Complete the translated sequence\n",
    "            self.mutant_aas[i] = str(new_translatable.translate())\n",
    "            \n",
    "    def get_readable_regions(self):\n",
    "        \n",
    "        # Get the read lengths of the forward and reverse reads\n",
    "        self.forward_readlength = self.read_length - ADAPTER_LENGTH_F - BARCODE_LENGTH\n",
    "        self.reverse_readlength = self.read_length - ADAPTER_LENGTH_R - BARCODE_LENGTH\n",
    "        \n",
    "        # Get the basepair readable positions\n",
    "        all_bp_positions = np.arange(len(self.ref_seq))\n",
    "        self.forward_readable_bps = set(all_bp_positions[:self.forward_readlength])\n",
    "        self.reverse_readable_bps = set(all_bp_positions[(self.ref_length - self.reverse_readlength):])\n",
    "        self.readable_bp_positions = self.forward_readable_bps.union(self.reverse_readable_bps)\n",
    "        \n",
    "        # Get the amino acid readable positions\n",
    "        self.readable_aa_positions = set(pos for pos, bps in self.codon_to_bp.items() \n",
    "                                         if all(bp in self.readable_bp_positions for bp in bps))\n",
    "        \n",
    "    def get_perfect_count_array(self):\n",
    "        \n",
    "        # Get the perfect counts for basepairs\n",
    "        self.perfect_counts = np.zeros(self.ref_length, dtype = int)\n",
    "        self.perfect_counts[:self.forward_readlength] += 1\n",
    "        self.perfect_counts[(self.ref_length - self.reverse_readlength):] += 1\n",
    "        \n",
    "        # Get the perfect counts for amino acids\n",
    "        n_aas = len(self.aa_ref_seq)\n",
    "        self.perfect_counts_aa = np.zeros(n_aas, dtype = int)\n",
    "\n",
    "        # Loop over the counts array and increment appropriately\n",
    "        for i in range(n_aas):\n",
    "            self.perfect_counts_aa[i] += np.min(self.perfect_counts[self.codon_to_bp[i]])\n",
    "        \n",
    "    # Identify readable variable positions\n",
    "    def find_variable_positions(self):\n",
    "        \n",
    "        # Get the readable variable basepair positions\n",
    "        self.readable_variable_bp = np.array([[i for i, (char1, char2) in \n",
    "                                               enumerate(zip(self.ref_seq, mutant_seq))\n",
    "                                               if (char1 != char2 and i in self.readable_bp_positions)]\n",
    "                                              for mutant_seq in self.mutant_seqs])\n",
    "        \n",
    "        # Find readable codons\n",
    "        mutated_codons_by_seq = [[i for i, (char1, char2) in enumerate(zip(self.aa_ref_seq, mutant_seq))\n",
    "                                 if (char1 != char2)] for mutant_seq in self.mutant_aas]\n",
    "                \n",
    "        self.readable_codons = [[codon for codon in seq if \n",
    "                                all(base_ind in self.readable_bp_positions for base_ind in self.codon_to_bp[codon])]\n",
    "                               for seq in mutated_codons_by_seq]\n",
    "        \n",
    "#     # Attach barcodes and adapters to mutant sequences\n",
    "#     def complete_sequences(self, FBC, RBC):\n",
    "        \n",
    "#         # Build complete sequences that can be saved as inputs\n",
    "#         self.finished_seqs_f = []\n",
    "#         self.finished_seqs_r = []\n",
    "#         for mutant_seq in self.mutant_seqs:\n",
    "            \n",
    "#             # Pull readable regions of the sequences\n",
    "#             forward_seq_unfinished = mutant_seq[:self.forward_readlength]\n",
    "#             rev_seq_unfinished = Seq(mutant_seq[(self.ref_length - self.reverse_readlength):])\n",
    "#             rev_seq_unfinished = str(rev_seq_unfinished.reverse_complement())\n",
    "\n",
    "#             # Finish the sequences and store\n",
    "#             self.finished_seqs_f.append(Seq(\"\".join((FBC, F_ADAPTER, forward_seq_unfinished))))\n",
    "#             self.finished_seqs_r.append(Seq(\"\".join((RBC, R_ADAPTER, rev_seq_unfinished))))    \n",
    "                \n",
    "    # Add noise to the completed sequences. Build the outputs (but don't save) as we go along.\n",
    "    def build_output(self, bp_q_cutoff, n_reads = None):\n",
    "        \n",
    "        # Define the allowed range of quality scores\n",
    "        high_q_range = np.arange(bp_q_cutoff, 40)\n",
    "        low_q_range = np.arange(1, bp_q_cutoff)\n",
    "\n",
    "        # Decide on the number of sequences in the well\n",
    "        if n_reads is None:\n",
    "            self.n_seqs_in_well = np.random.randint(*allowed_n_reads)\n",
    "        else:\n",
    "            self.n_seqs_in_well\n",
    "\n",
    "        # Get the number of times each sequence will be represented\n",
    "        n_reps = self.n_seqs_in_well // self.n_variants\n",
    "        remainder = self.n_seqs_in_well % self.n_variants\n",
    "        total_counts_by_seq = np.full(self.n_variants, n_reps)\n",
    "        total_counts_by_seq[0] += remainder\n",
    "        \n",
    "        # Loop over each sequence\n",
    "        for n_seqs, bp_seq, aa_seq in zip(total_counts_by_seq, self.mutant_seqs, self.mutant_aas):\n",
    "            \n",
    "            # Loop over the number of sequences\n",
    "            for _ in range(n_seqs):\n",
    "                \n",
    "                # Create booleans \n",
    "                ruin_f_indel = False\n",
    "                ruin_r_indel = False\n",
    "                ruin_f_qual = False\n",
    "                ruin_r_qual = False\n",
    "            \n",
    "                # Generate a count array for the sequence\n",
    "                seq_count_array = self.perfect_counts.copy()\n",
    "                aa_count_array = self.perfect_counts_aa.copy()\n",
    "\n",
    "                # Make a list for storing the new forward and reverse sequences\n",
    "                new_f = []\n",
    "                new_r = []\n",
    "\n",
    "                # Decide if we want to ruin the forward sequence with an insertion or deletion\n",
    "                if np.random.rand() < indel_freq:\n",
    "                    ruin_f_indel = True\n",
    "                    \n",
    "                # Decide if we want to ruin the reverse sequence with an insertion or deletion\n",
    "                if np.random.rand() < indel_freq:\n",
    "                    ruin_r_indel = True\n",
    "\n",
    "                # Decide if we want to ruin the forward sequence by dropping its quality\n",
    "                if np.random.rand() < low_qual_freq:\n",
    "                    ruin_f_qual = True\n",
    "                    \n",
    "                # Decide if we want to ruin the reverse sequence by dropping its quality\n",
    "                if np.random.rand() < low_qual_freq:\n",
    "                    ruin_r_qual = True\n",
    "\n",
    "                # If the sequence is ruined by either of the above, subtract 1 from the appropriate\n",
    "                # region of the count array. Note that this read won't count toward combinations.\n",
    "                \n",
    "                \n",
    "                \n",
    "                # Loop over each basepair in the sequence\n",
    "                for bp in bp_seq:\n",
    "\n",
    "                    # If this bp is not in a readable region, assign random \n",
    "\n",
    "                    # If the forward and reverse regions overlap, decide if we want a disagreement in \n",
    "                    # base called. If there is a disagreement, subtract 1 from the appropriate position\n",
    "                    # of the count array. The reverse read will be used if there is a tie in qualities. If both \n",
    "                    # reads are less than the q-score threshold, subtract 2 from the region of the count\n",
    "                    # array and note that this read won't count toward combinations.\n",
    "                    # If only 1 read is less than the q-score threshold, subtract 1; the read should still \n",
    "                    # count toward combinations.\n",
    "\n",
    "                    # If we are in non-overlapping regions and the quality is low, subtract 1 from the\n",
    "                    # appropriate count position and note that this read won't count toward combinations\n",
    "                \n",
    "        \n",
    "        \n",
    "    # Attach barcodes and adapters to mutant sequences\n",
    "    def complete_sequences(self, FBC, RBC):\n",
    "        \n",
    "        # Build complete sequences that can be saved as inputs\n",
    "        self.finished_seqs_f = []\n",
    "        self.finished_seqs_r = []\n",
    "        for mutant_seq in self.mutant_seqs:\n",
    "            \n",
    "            # Pull readable regions of the sequences\n",
    "            forward_seq_unfinished = mutant_seq[:self.forward_readlength]\n",
    "            rev_seq_unfinished = Seq(mutant_seq[(self.ref_length - self.reverse_readlength):])\n",
    "            rev_seq_unfinished = str(rev_seq_unfinished.reverse_complement())\n",
    "\n",
    "            # Finish the sequences and store\n",
    "            self.finished_seqs_f.append(Seq(\"\".join((FBC, F_ADAPTER, forward_seq_unfinished))))\n",
    "            self.finished_seqs_r.append(Seq(\"\".join((RBC, R_ADAPTER, rev_seq_unfinished))))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = WellContents()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OutputData():\n",
    "    \n",
    "    # Initialize\n",
    "    def __init__(self, n_seqs, perfect_counts, perfect_counts_aa):\n",
    "        \n",
    "        # Create lists for storing sequences\n",
    "        self.f_seqs = [None] * n_seqs\n",
    "        self.r_seqs = [None] * n_seqs\n",
    "        \n",
    "        # Create counters to store relevant results\n",
    "        self.passing_bp_combos = n_seqs\n",
    "        self.passing_aa_combos = n_seqs\n",
    "        self.bp_counts_by_pos = perfect_counts * n_seqs\n",
    "        self.aa_counts_by_pos = perfect_counts_aa * n_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decide_on_ruin():\n",
    "    \n",
    "    # Create booleans \n",
    "    ruin_f_indel = True if np.random.rand() < indel_freq else False\n",
    "    ruin_r_indel = True if np.random.rand() < indel_freq else False\n",
    "    ruin_f_qual = True if np.random.rand() < low_qual_freq else False\n",
    "    ruin_r_qual = True if np.random.rand() < low_qual_freq else False\n",
    "    \n",
    "    return ruin_f_indel, ruin_r_indel, ruin_f_qual, ruin_r_qual\n",
    "\n",
    "def enact_ruin(seq, qualities, q_thresh):\n",
    "    \n",
    "    # Create booleans to see if we will ruin a sequence by adding an\n",
    "    # insertion or deletion, dropping its quality, or making it too short\n",
    "    ruin_indel = True if np.random.rand() < indel_freq else False\n",
    "    ruin_qual = True if np.random.rand() < low_qual_freq else False\n",
    "    ruin_length = True if np.random.rand() < low_length_freq else False\n",
    "    \n",
    "    # If we aren't ruining anything, just return the sequence with a \n",
    "    # quality score\n",
    "    if not any(ruin_indel, ruin_qual):\n",
    "        return seq, qualities\n",
    "    \n",
    "    # Create a new sequence array\n",
    "    seq = list(seq)\n",
    "    \n",
    "    # If we are ruining something, decide on fate\n",
    "    if ruin_indel:\n",
    "        \n",
    "        # Choose a point for insertion or deletion\n",
    "        indel_point = np.random.randint(2, len(seq) - 3)\n",
    "        \n",
    "        # Route for insertion\n",
    "        if np.random.choice([True, False]):\n",
    "            \n",
    "            # Choose new quality and base\n",
    "            new_qual = np.random.choice(1, 40)\n",
    "            new_base = np.random.choice(allowed_bases_tup)\n",
    "            \n",
    "            # Update the sequence\n",
    "            seq = list(seq)\n",
    "            seq.insert(indel_point, new_base)\n",
    "            seq = \"\".join(seq)\n",
    "            \n",
    "            # Update the qualities\n",
    "            qualities = np.insert(qualities, indel_point)\n",
    "            \n",
    "        # Route for deletion\n",
    "        else:\n",
    "            \n",
    "            # Update the sequence\n",
    "            del(seq[indel_points])\n",
    "            seq = \"\".join(seq)\n",
    "            \n",
    "            # Update the qualities\n",
    "            qualities = np.delete(qualities, indel_point)\n",
    "    \n",
    "    # If we are ruining by quality, do this here\n",
    "    if ruin_qual:\n",
    "        \n",
    "        # Take random qualities below the threshold\n",
    "        qualities = np.random.randint()\n",
    "        \n",
    "    \n",
    "\n",
    "# Define the allowed range of quality scores\n",
    "high_q_range = np.arange(bp_q_cutoff, 40)\n",
    "low_q_range = np.arange(1, bp_q_cutoff)\n",
    "\n",
    "# Decide on the number of sequences in the well\n",
    "if n_reads is None:\n",
    "    test.n_seqs_in_well = np.random.randint(*allowed_n_reads)\n",
    "else:\n",
    "    test.n_seqs_in_well\n",
    "\n",
    "# Get the number of times each sequence will be represented\n",
    "n_reps = test.n_seqs_in_well // test.n_variants\n",
    "remainder = test.n_seqs_in_well % test.n_variants\n",
    "total_counts_by_seq = np.full(test.n_variants, n_reps)\n",
    "total_counts_by_seq[0] += remainder\n",
    "\n",
    "# Create a list for storing all results\n",
    "all_results = []\n",
    "\n",
    "# Loop over each sequence\n",
    "for n_seqs, bp_seq, aa_seq in zip(total_counts_by_seq, test.mutant_seqs, test.mutant_aas):\n",
    "    \n",
    "    # Create a container for storing output data\n",
    "    output_container = OutputData()\n",
    "    \n",
    "    # Loop over the number of sequences\n",
    "    for _ in range(n_seqs):\n",
    "\n",
    "\n",
    "\n",
    "        # If the sequence is ruined by either of the above, subtract 1 from the appropriate\n",
    "        # region of the count array. Note that this read won't count toward combinations.\n",
    "\n",
    "\n",
    "\n",
    "        # Loop over each basepair in the sequence\n",
    "        for bp in bp_seq:\n",
    "\n",
    "            # Randomly terminate early to have some sequences that are shorter than expected (but\n",
    "            # that still will work)\n",
    "            \n",
    "            # If this bp is not in a readable region, assign random \n",
    "\n",
    "            # If the forward and reverse regions overlap, decide if we want a disagreement in \n",
    "            # base called. If there is a disagreement, subtract 1 from the appropriate position\n",
    "            # of the count array. The reverse read will be used if there is a tie in qualities. If both \n",
    "            # reads are less than the q-score threshold, subtract 2 from the region of the count\n",
    "            # array and note that this read won't count toward combinations.\n",
    "            # If only 1 read is less than the q-score threshold, subtract 1; the read should still \n",
    "            # count toward combinations.\n",
    "\n",
    "            # If we are in non-overlapping regions and the quality is low, subtract 1 from the\n",
    "            # appropriate count position and note that this read won't count toward combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
       "       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
       "       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
       "       1])"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perfect_counts_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TGCATGTATATTTCATTGCACAGAATAAGCCACCACACGTCCGTAACACACCCGCCCTTCGCCGAGGAGTTAGTTCGGAATCGAGTTGCCTTTTTGATTGCGCGCTGTCCTCTAAGCCCGCGGCAGTTATAGCTGCGACCGTACAGTACATCATATTGGCGTATACGAACGGCTGATGGATGGACTCCGATCTCCGTTCCATTTAGGTGTCTAAATAAATTGAAACATCCAATATCCGATCGATTTTGTACAAGTGGTGTAATACGCTGCGATCCTTCATGGTCTCCGTGCTGGCCACGGGCGCGTTTTCTATCTGAATACCTCGCGTGATCTAGGATTACTATGAACCACCGTCAGCGGATCGTGGTTGCCAGTTGATAGCTTGCCACTAATTCCTTGA']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.mutant_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Decide if we want to eliminate the forward or reverse read in this count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start by calculating and testing outputs if no noise is added. Add noise later. Then you just adjust the ideal output based on the noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_output_records(average_q_cutoff = 25, bp_q_cutoff = 30,\n",
    "                         variable_count = 10, read_length = 150,\n",
    "                        ref_length = None):\n",
    "    pass\n",
    "\n",
    "average_q_cutoff = 25\n",
    "bp_q_cutoff = 30\n",
    "variable_count = 10\n",
    "read_length = 150\n",
    "ref_length = None\n",
    "\n",
    "np.random.seed(2)\n",
    "\n",
    "# Create a counter to index the input sequences. Create lists to\n",
    "# store all outputs\n",
    "output_counter = 0\n",
    "all_out_f = []\n",
    "all_out_r = []\n",
    "\n",
    "# Build a list for storing the reference sequences\n",
    "output_cols = [\"PlateName\", \"IndexPlate\", \"Well\", \"ReferenceSequence\",\n",
    "              \"InFrameBase\", \"BpIndStart\", \"AaIndStart\"]\n",
    "ref_seq_file_array = [None] * len(barcode_df)\n",
    "wells = ref_seq_file_array.copy()\n",
    "\n",
    "# Loop over the barcode plate\n",
    "for barcode_ind, row in enumerate(barcode_df.itertuples(index = False)):\n",
    "        \n",
    "    # Create a well with specific content\n",
    "    well = WellContents(ref_length = ref_length, read_length = read_length)   \n",
    "    wells[barcode_ind] = well\n",
    "    \n",
    "    # Populate the reference sequence row\n",
    "    ref_seq_file_array[barcode_ind] = [row.IndexPlate + \"_TEST\", row.IndexPlate, row.Well,\n",
    "                                      well.ref_seq, well.in_frame + 1, 1232, 23]\n",
    "    \n",
    "    # Define the allowed range of quality scores\n",
    "    high_q_range = np.arange(bp_q_cutoff, 40)\n",
    "\n",
    "    # Decide on the number of sequences in the well\n",
    "    n_seqs_in_well = np.random.randint(*allowed_n_reads)\n",
    "\n",
    "    # Get the number of times each sequence will be represented\n",
    "    n_reps = n_seqs_in_well // well.n_variants\n",
    "    remainder = n_seqs_in_well % well.n_variants\n",
    "    total_counts_by_seq = np.full(well.n_variants, n_reps)\n",
    "    total_counts_by_seq[0] += remainder\n",
    "\n",
    "    # Loop over the sequences and build both the expected output\n",
    "    # and the desired input\n",
    "    output_records_f = [None] * n_seqs_in_well\n",
    "    output_records_r = output_records_f.copy()\n",
    "    temp_counter = 0\n",
    "    for total_counts, mutant_seq, mutant_aa_seq in zip(total_counts_by_seq, well.mutant_seqs, well.mutant_aas):\n",
    "        \n",
    "        # Build outputs\n",
    "        for _ in range(total_counts):\n",
    "\n",
    "            # Pull the forward sequence\n",
    "            forward_seq_unfinished = mutant_seq[:well.forward_readlength]\n",
    "            rev_seq_unfinished = Seq(mutant_seq[(well.ref_length - well.reverse_readlength):])\n",
    "            rev_seq_unfinished = str(rev_seq_unfinished.reverse_complement())\n",
    "\n",
    "            # Finish the sequences\n",
    "            finished_f = Seq(\"\".join((row.FBC, F_ADAPTER, forward_seq_unfinished)))\n",
    "            finished_r = Seq(\"\".join((row.RBC, R_ADAPTER, rev_seq_unfinished)))\n",
    "\n",
    "            # Create sequence records\n",
    "            output_records_f[temp_counter] = SeqRecord(finished_f,\n",
    "                                                       id = id_template.substitute(num = output_counter),\n",
    "                                                       letter_annotations = {\"phred_quality\": np.random.choice(high_q_range, len(finished_f))})\n",
    "            output_records_r[temp_counter] = SeqRecord(finished_r,\n",
    "                                                       id = id_template.substitute(num = output_counter),\n",
    "                                                       letter_annotations = {\"phred_quality\": np.random.choice(high_q_range, len(finished_r))})\n",
    "\n",
    "            # Update counters\n",
    "            output_counter += 1 \n",
    "            temp_counter += 1\n",
    "            \n",
    "    # Extend the overall records\n",
    "    all_out_f.extend(output_records_f)\n",
    "    all_out_r.extend(output_records_r)            \n",
    "    \n",
    "# Save results after shuffling order of output records\n",
    "shuffle(all_out_f)\n",
    "shuffle(all_out_r)\n",
    "with open(\"deSeqValidation/Test_R1_001.fastq\", \"w\") as f:\n",
    "    SeqIO.write(all_out_f, f, \"fastq\")\n",
    "with open(\"deSeqValidation/Test_R2_001.fastq\", \"w\") as f:\n",
    "    SeqIO.write(all_out_r, f, \"fastq\")\n",
    "output_df = pd.DataFrame(ref_seq_file_array, columns = output_cols)\n",
    "output_df.to_csv(\"./deSeqValidation/TestRefSeq.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(wells[6].codon_mutation_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f_bc, r_bc = barcode_df.loc[(barcode_df.IndexPlate == \"DI01\")&(barcode_df.Well == \"A01\"), [\"FBC\", \"RBC\"]].values.flatten()\n",
    "id_counter = 0\n",
    "\n",
    "\n",
    "    \n",
    "#     # Grab the mutated bases\n",
    "#     mutant_bases = [[base_ind, orig_base, new_base] for \n",
    "#                     base_ind, (orig_base, new_base) in\n",
    "#                     enumerate(zip(test.ref_seq, mutant_seq))\n",
    "#                    if orig_base != new_base]\n",
    "    \n",
    "#     # Grab the mutated amino acids\n",
    "#     mutant_aas = [[aa_ind, orig_aa, new_aa] for \n",
    "#                  aa_ind, (orig_aa, new_aa) in\n",
    "#                  enumerate(zip(test.aa_ref_seq, mutant_aa_seq))\n",
    "#                  if orig_aa != new_aa]\n",
    "    \n",
    "#     # Grab the expected counts for the bases\n",
    "#     expected_base_counts = [test.perfect_counts[mutant_ind] for mutant_ind, _, _ in mutant_bases]\n",
    "    \n",
    "#     Get the expected counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(output_records_f[0].letter_annotations[\"phred_quality\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_records_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Loop over each variable sequence\n",
    "\n",
    "\n",
    "# Create a counter for the number of sequences in a well\n",
    "\n",
    "\n",
    "\n",
    "# Determine codons with poor quality bases that we mutated.\n",
    "# bad_base_test = np.random.rand(len(test.codon_mutation_positions)) < bad_base_freq\n",
    "# codons_with_bad_base = test.codon_mutation_positions[bad_base_test]\n",
    "\n",
    "\n",
    "# Determine the bad bases for the variable each sequence.\n",
    "# bad_base_checks = np.random.rand(n_seqs_in_well, test.ref_length) < bad_base_freq\n",
    "\n",
    "# Decide if there are any bad bases in the sequences\n",
    "# bad_base_checker = np.random.rand(n_seqs_in_well) < bad_base_in_seq_freq\n",
    "\n",
    "# Loop over the bad base checker and decide (a) the number of bad\n",
    "# positions and (b) where those positions are. \n",
    "\n",
    "# After sequences are split, decide if we want to tank just one of them (e.g.\n",
    "# by adding insertions or deletions or dropping the q-score very low)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_seqs_in_well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_counts_by_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons_with_bad_base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the overall number of reads that we want for the well\n",
    "total_possible_reads = np.random.randint(*allowed_n_reads)\n",
    "\n",
    "# Decide if the well is dead. Evaluate possible conditions to make it dead.\n",
    "if random.rand() < dead_well_freq:\n",
    "    \n",
    "    # Choose \"1\", \"2\", or \"3\". \"1\" means make no usable reads (either by insertion, deletion\n",
    "    # or low average read quality)\n",
    "\n",
    "# Decide if we want this to be wild type \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a kill reads function that either makes an insertion/deletion\n",
    "# or else just wipes out functionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we need a function that...\n",
    "# 1. Determines the overall count of each read\n",
    "# 2. Assigns a target frequency for each position\n",
    "# 3. Assigns a target count for each position\n",
    "# 4. Breaks into forward and reverse reads\n",
    "# 5. Loops over reads and assigns quality scores to the different bases\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "build_reference_and_variants()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python379jvsc74a57bd09f9217248950819e3e330b05aa6a64b63adb55d52a2073ac26daa1eea072b080",
   "display_name": "Python 3.7.9 64-bit ('deSeq': conda)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}