{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook serves as a development ground for the new alignment protocol. Note that the clustal omega version being used was downloaded on 10/26/2020 from \"http://www.clustal.org/omega/\" and was renamed to \"clustalo\" from the binary file clustalo-1.2.4-Ubuntu-x86_64."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load necessary modules\n",
    "from Bio import SeqIO\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "BarcodeLength = 7\n",
    "AdapterLengthF = 27\n",
    "AdapterLengthR = 26\n",
    "\n",
    "# Define an object that holds BioPython SeqRecords\n",
    "class SeqPair():\n",
    "    \n",
    "    # Parse forward reads during initialization\n",
    "    def __init__(self, f_record):\n",
    "        \n",
    "        # Assign the forward barcode and the adapterless sequence\n",
    "        self._f_barcode = str(f_record.seq)[:BarcodeLength]\n",
    "        self._f_adapterless = f_record[(BarcodeLength + AdapterLengthF):]\n",
    "        self._f_len = len(f_record)\n",
    "        \n",
    "        # Until we have a reverse record paired, this is an orpha\n",
    "        self._orphan = True\n",
    "    \n",
    "    # Assign a paired reverse read\n",
    "    def assign_partner(self, r_record):\n",
    "        \n",
    "        # Assign the reverse barcode and adapterless sequence. \n",
    "        # We want to have the reverse complement of this sequence to match\n",
    "        # the reference sequence\n",
    "        self._r_barcode = str(r_record.seq)[:BarcodeLength]\n",
    "        sliced_r = r_record[(BarcodeLength + AdapterLengthR):]\n",
    "        self._r_adapterless = sliced_r.reverse_complement()\n",
    "        self._r_len = len(r_record)\n",
    "        \n",
    "        # This is no longer an orphan\n",
    "        self._orphan = False\n",
    "        \n",
    "    # Make all the properties\n",
    "    @property\n",
    "    def f_barcode(self):\n",
    "        return self._f_barcode\n",
    "    \n",
    "    @property\n",
    "    def f_adapterless(self):\n",
    "        return self._f_adapterless\n",
    "    \n",
    "    @property\n",
    "    def f_len(self):\n",
    "        return self._f_len\n",
    "    \n",
    "    @property\n",
    "    def r_barcode(self):\n",
    "        return self._r_barcode\n",
    "    \n",
    "    @property\n",
    "    def r_adapterless(self):\n",
    "        return self._r_adapterless\n",
    "    \n",
    "    @property\n",
    "    def r_len(self):\n",
    "        return self._r_len\n",
    "    \n",
    "    @property\n",
    "    def orphan(self):\n",
    "        return self._orphan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write a function for loading and pairing fastq files\n",
    "def load_fastq(f_loc, r_loc):\n",
    "\n",
    "    # Create a dictionary that links id to sequence object\n",
    "    id_to_reads = {}\n",
    "    print(\"Loading forward reads...\")\n",
    "    all_f_recs = list(SeqIO.parse(f_loc, \"fastq\"))\n",
    "    id_to_reads = {f_record.id: SeqPair(f_record) for f_record in all_f_recs}\n",
    "    \n",
    "    # Associate reverse reads with the forward\n",
    "    print(\"Loading reverse reads...\")\n",
    "    all_r_recs = list(SeqIO.parse(r_loc, \"fastq\"))\n",
    "    for r_record in all_r_recs:\n",
    "\n",
    "        # If there is no partern in id_to_reads, continue\n",
    "        if r_record.id not in id_to_reads:\n",
    "            continue\n",
    "\n",
    "        # Otherwise, attach the reverse record\n",
    "        else:\n",
    "            id_to_reads[r_record.id].assign_partner(r_record)\n",
    "            \n",
    "    # Only keep records that have a partner\n",
    "    return tuple([seqpair for seqpair in id_to_reads.values() if not seqpair.orphan])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading forward reads...\n",
      "Loading reverse reads...\n"
     ]
    }
   ],
   "source": [
    "# Load fastq files\n",
    "all_seqpairs = load_fastq(\"./TestData/20200205_ssSeq/CHL2_S199_L001_R1_001.fastq\",\n",
    "                          \"./TestData/20200205_ssSeq/CHL2_S199_L001_R2_001.fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLOWED_WELLS = {'A01', 'A02', 'A03', 'A04', 'A05', 'A06', 'A07', 'A08', 'A09', \n",
    "                 'A10', 'A11', 'A12', 'B01', 'B02', 'B03', 'B04', 'B05', 'B06', \n",
    "                 'B07', 'B08', 'B09', 'B10', 'B11', 'B12', 'C01', 'C02', 'C03', \n",
    "                 'C04', 'C05', 'C06', 'C07', 'C08', 'C09', 'C10', 'C11', 'C12', \n",
    "                 'D01', 'D02', 'D03', 'D04', 'D05', 'D06', 'D07', 'D08', 'D09', \n",
    "                 'D10', 'D11', 'D12', 'E01', 'E02', 'E03', 'E04', 'E05', 'E06', \n",
    "                 'E07', 'E08', 'E09', 'E10', 'E11', 'E12', 'F01', 'F02', 'F03', \n",
    "                 'F04', 'F05', 'F06', 'F07', 'F08', 'F09', 'F10', 'F11', 'F12', \n",
    "                 'G01', 'G02', 'G03', 'G04', 'G05', 'G06', 'G07', 'G08', 'G09', \n",
    "                 'G10', 'G11', 'G12', 'H01', 'H02', 'H03', 'H04', 'H05', 'H06', \n",
    "                 'H07', 'H08', 'H09', 'H10', 'H11', 'H12'}\n",
    "\n",
    "# Load the index map and reference sequencefile\n",
    "index_map = pd.read_csv(\"/home/brucejwittmann/GitRepos/ssSeq/ssSeqSupport/IndexMap.csv\")\n",
    "ref_seq_crude = pd.read_csv(\"/home/brucejwittmann/GitRepos/ssSeq/AlignmentDev/TestData/20200205_ssSeq/RefSeqs.csv\")\n",
    "\n",
    "# Expand each reference sequence\n",
    "updated_ref_array = []\n",
    "for row in ref_seq_crude.itertuples(index = False):\n",
    "    updated_ref_array.extend([[row.PlateName, row.IndexPlate, well, row.ReferenceSequence]\n",
    "                             for well in ALLOWED_WELLS])\n",
    "    \n",
    "# Define the complete reference sequence dataframe\n",
    "complete_ref_seq = pd.DataFrame(updated_ref_array, columns = (\"PlateName\", \"IndexPlate\", \"Well\", \"ReferenceSequence\"))\n",
    "\n",
    "# Join on plate and well\n",
    "merged_dfs = complete_ref_seq.merge(index_map, on = (\"IndexPlate\", \"Well\"))\n",
    "\n",
    "# Map barcode to reference sequence, plate, and well\n",
    "bc_to_ref_plate_well = {(row.FBC, row.RBC): {\"IndexPlate\": row.IndexPlate,\n",
    "                                             \"PlateNickname\": row.PlateName,\n",
    "                                             \"Well\": row.Well,\n",
    "                                             \"ReferenceSequence\": row.ReferenceSequence}\n",
    "                       for row in merged_dfs.itertuples(index = False)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Well():\n",
    "    \n",
    "    # Initialization assigns attributes, reference sequences, and sequence pairs\n",
    "    def __init__(self, seqpairs, refseq_df_info):\n",
    "        \n",
    "        # Assign the sequence pairs as an attribute and unpack the refseq info\n",
    "        self._seqpairs = seqpairs\n",
    "        self._index_plate = refseq_df_info[\"IndexPlate\"]\n",
    "        self._plate_nickname = refseq_df_info[\"PlateNickname\"]\n",
    "        self._well = refseq_df_info[\"Well\"]\n",
    "        self._reference_sequence = refseq_df_info[\"ReferenceSequence\"]\n",
    "        \n",
    "    # Write a function that outputs fasta files for all of the seqpairs\n",
    "    def write_fastas(self):\n",
    "        pass\n",
    "    \n",
    "    # Write a function that uses Clustal Omega to make an MSA of the alignment\n",
    "    def align(self):\n",
    "        pass\n",
    "    \n",
    "    # Write a function that analyzes the alignment output and identifies variable\n",
    "    # positions\n",
    "    def find_variable_positions(self, first_in_frame = None):\n",
    "        \n",
    "        # Error if first in frame is not provided\n",
    "        if first_in_frame is None:\n",
    "            raise \n",
    "        \n",
    "    # Define properties\n",
    "    @property\n",
    "    def seqpairs(self):\n",
    "        return self._seqpairs\n",
    "    \n",
    "    @property\n",
    "    def index_plate(self):\n",
    "        return self._index_plate\n",
    "    \n",
    "    @property\n",
    "    def plate_nickname(self):\n",
    "        return self._plate_nickname\n",
    "    \n",
    "    @property\n",
    "    def well(self):\n",
    "        return self._well\n",
    "    \n",
    "    @property\n",
    "    def reference_sequence(self):\n",
    "        return self._reference_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_seqpairs_to_well(all_seqpairs, bc_to_ref_plate_well):\n",
    "\n",
    "    # Loop over all seqpairs and assign to wells\n",
    "    print(\"Assigning sequences to wells...\")\n",
    "    well_pairs = {}\n",
    "    for pair in all_seqpairs:\n",
    "\n",
    "        # Grab the well ID and see if it is a real well. Continue\n",
    "        # if it is not. \"Fake\" wells are those that result from \n",
    "        # sequencing errors\n",
    "        well_id = (pair.f_barcode, pair.r_barcode)\n",
    "        if well_id not in bc_to_ref_plate_well:\n",
    "            continue\n",
    "        \n",
    "        # Check to see if we have seen this well already.\n",
    "        # If we have seen it, append to growing list. If we have not,\n",
    "        # start a new list\n",
    "        if well_id in well_pairs:\n",
    "            well_pairs[well_id].append(pair)\n",
    "        else:\n",
    "            well_pairs[well_id] = [pair]\n",
    "            \n",
    "    # Now build the well object\n",
    "            \n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
